### PR TITLE
feat: remove font family hashing in next/font css

### PR DESCRIPTION
**Pull request by blvdmitry**
### What?

Adding support for supporting a custom fontFamily name when using next/font

### Why?

By default, next/font hashes the font name when generating css to achieve proper scoping. 
However, that makes it impossible to use next/font with 3rd party libraries that provide CSS with pre-defined font names.

### How?

To solve this, I've added a new argument to the next/font function call – `usedFontFamilyName`.
It allows developers to pick the fontFamily name that is going to be used in the CSS output instead of the default one and make it work with vendor CSS files.

```
import { Inter } from "next/font/google";

const inter = Inter({
  subsets: ["latin"],
  fixedFontFamily: "Inter",
});
```

Fixes [#43452](https://github.com/vercel/next.js/discussions/43452)

---

Edit:

I've changed the implementation to use `disabledFontFamilyHashing` boolean flag which removes the hashing but keeps the original font family name instead of allowing a custom name

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 061f77f6384f0af43de28f7be3f996df7a3547dc

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 061f77f6384f0af43de28f7be3f996df7a3547dc

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by blvdmitry** [comment]
Not exactly sure why the mdx change is failing exactly since the headings and link approach is exactly the same as for other links in the arguments table. Looking into why this might be happening but would appreciate the help in case you have more context about this.

**Comment by blvdmitry** [comment]
Fixed the documentation validation

**Comment by filiptdz** [comment]
Great work! I suggest checking for `declarations: [{ prop: 'font-family', value: 'New Name' }]` to use as the font family name. If this is too messy, maybe you could rename `usedFontFamilyName` to `usedFontFamily` or `fixedFontFamily` as it'd bring it closer to the CSS property name

**Comment by blvdmitry** [comment]
In the discussion someone had issues using the declarations approach but agree that name could be shorter. I'll go through other options again once I'm back from vacation :)

**Comment by maccman** [comment]
Excited for this!

**Comment by blvdmitry** [comment]
Updated to `fixedFontFamily`

**Comment by blvdmitry** [comment]
@shuding @ijjk any though on the PR? It doesn't look exactly like a very urgent feature but it was requested by a bunch of people and helps a lot when using next/font with 3rd party libraries. 

Anything you would expect from me to get this going further?

**Comment by yashsway** [comment]
> Great work! I suggest checking for `declarations: [{ prop: 'font-family', value: 'New Name' }]` to use as the font family name. If this is too messy, maybe you could rename `usedFontFamilyName` to `usedFontFamily` or `fixedFontFamily` as it'd bring it closer to the CSS property name

I had issues using the `declarations` approach. Made sense in the docs but couldn't use it: Next after building threw an error mentioning that "font-family" isn't a valid `prop` that be used in the `declarations` key. Strange because it seemed like you could override any property available in `@font-face` other than `src` I guess. Or maybe it wasn't clear in the docs as to what the restrictions were.

**Comment by blvdmitry** [comment]
I can definitely resolve any conflicts but I parked it for now until there is any response from the Next.js team who can enable the rest of the CI. Most of the conflicts in this PR usually happen because fonts metadata gets updated in the main branch so it just requires me to run one script.

**Comment by dqii** [comment]
Would love to see this PR merged!

**Comment by blvdmitry** [comment]
It's 2024 now, is there a chance we'll get a review from someone? 😄

**Review by sahilsarpalguru — APPROVED** [review]

**Comment by IgnusG** [comment]
@sahilsarpalguru good try 😅 But I fear we need a review from one of the codeowners like @timneutkens, @ijjk, @leerob or @wyattjoh to tag a few to bring this PR forward

**Comment by sahilsarpalguru** [comment]
> tag a few (my apologies for the noise) to bri

I want the codeowners to bring the attention to this PR, I recently got blocked in this scenerios. @IgnusG btw do we have any workaround to solve this ?

**Comment by blvdmitry** [comment]
> > tag a few (my apologies for the noise) to bri
> 
> I want the codeowners to bring the attention to this PR, I recently got blocked in this scenerios. @IgnusG btw do we have any workaround to solve this ?

My workaround right now is to load the fonts manually :)

**Comment by IgnusG** [comment]
> > tag a few (my apologies for the noise) to bri
> 
> I want the codeowners to bring the attention to this PR, I recently got blocked in this scenerios. @IgnusG btw do we have any workaround to solve this ?

Unfortunately none that is easy enough to integrate. What I do is to specify a CSS variable for the font through and then I use a small script to replace the static font name (eg. `"Inter"`) in my theme with the CSS variable. Of course this doesn't work if you can't modify your theme/font CSS definitions 😞

```typescript
const fontInter = Inter({
  subsets: ['latin'],
  display: 'swap',
  variable: '--token-font-inter',
  fallback: ['Helvetica', 'Arial', 'sans-serif'],
});

const fontInterDefinition = {
  name: 'Inter',
  variable: '--token-font-inter',
};

function replaceFontFamily(
  fontFamily: string,
  nextFonts: { name: string; variable: string }[],
) {
  return nextFonts.reduce((result, font) => {
    return result.replaceAll(
      new RegExp(`['"]?${font.name}['"]?`, 'g'),
      `var(${font.variable})`,
    );
  }, fontFamily);
}
```

This PR would simplify this to:

```typescript
const fontInter = Inter({
  fixedFontName: 'Inter',
  subsets: ['latin'],
  display: 'swap',
  fallback: ['Helvetica', 'Arial', 'sans-serif'],
});
```

**Comment by blvdmitry** [comment]
Hey @ijjk, I haven't rebased it for a while since it was getting stale very quickly without being able to run the whole pipeline. Do you know if you or someone from the team will have time for a review, so I can rebase the branch and run all CI jobs?

**Review by Tarektouati — APPROVED** [review]

**Comment by blvdmitry** [comment]
@mahmud009 still waiting for someone with permissions to click that checkbox that unblocks the rest of the pipelines. After we get someone notice this MR – I'll rebase it again

**Comment by IgnusG** [comment]
Aghhh Next.js has been somewhat lacking in the community handling ever since the major 13 release 😞 

<details>
<summary>Rant rant rant - ignore if uninterested</summary>
This issue has been open for 6 months despite many pings to many maintainers with no replies. There's another major discussion going on regarding node API support in edge runtime which after a few maintainer comments in the beginning now is completely ignored too with only pings every couple of days begging for input.

Just the other day a blog post on an App router migration came out that identified major deficiencies in the way Next.js handles their feature development. Left and right everywhere I look people just suggest remix and building more and more adapters to simplify migrations over to it. It's like the framework is slowly eating itself alive or perhaps Vercel bit more than it can chew with the whole turbo effort 😢 
</details>

**Comment by samcx** [comment]
I will be taking a look at this!

**Comment by blvdmitry** [comment]
Thanks @samcx, I will rebase today in the evening then since I have to regenerate the available google fonts every now and then. Let me know if there are any other things you would need to be updated along the way 🙌

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| buildDuration | 11.8s | 11.8s | N/A |
| buildDurationCached | 6s | 5s | N/A |
| nodeModulesSize | 200 MB | 200 MB | ⚠️ +97.7 kB |
| nextStartRea..uration (ms) | 427ms | 429ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| 3f784ff6-HASH.js gzip | 53.4 kB | 53.4 kB | ✓ |
| 423.HASH.js gzip | 185 B | 181 B | N/A |
| 68-HASH.js gzip | 29.8 kB | 29.8 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 237 B | 241 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 100 kB | 100 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 502 B | 501 B | N/A |
| css-HASH.js gzip | 320 B | 322 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 256 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.18 kB | 4.18 kB | N/A |
| index-HASH.js gzip | 257 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | N/A |
| routerDirect..HASH.js gzip | 310 B | 311 B | N/A |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 306 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 106 B | 106 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| _buildManifest.js gzip | 482 B | 482 B | ✓ |
| Overall change | 482 B | 482 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| index.html gzip | 529 B | 526 B | N/A |
| link.html gzip | 541 B | 539 B | N/A |
| withRouter.html gzip | 523 B | 521 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94 kB | 94.1 kB | N/A |
| page.js gzip | 150 kB | 150 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 622 B | N/A |
| middleware-r..fest.js gzip | 151 B | 149 B | N/A |
| middleware.js gzip | 37.6 kB | 37.6 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | blvdmitry/next.js next-font-fixed-name | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 95.8 kB | 95.8 kB | ✓ |
| app-page-tur..prod.js gzip | 96.4 kB | 96.4 kB | ✓ |
| app-page-tur..prod.js gzip | 91 kB | 91 kB | ✓ |
| app-page.run...dev.js gzip | 142 kB | 142 kB | ✓ |
| app-page.run..prod.js gzip | 90.3 kB | 90.3 kB | ✓ |
| app-route-ex...dev.js gzip | 22.2 kB | 22.2 kB | ✓ |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.5 kB | 14.5 kB | ✓ |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | ✓ |
| app-route.ru..prod.js gzip | 14.5 kB | 14.5 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.7 kB | 9.7 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.7 kB | 22.7 kB | ✓ |
| pages.runtim..prod.js gzip | 22 kB | 22 kB | ✓ |
| server.runti..prod.js gzip | 49.7 kB | 49.7 kB | ✓ |
| Overall change | 933 kB | 933 kB | ✓ |

</details>

</details>
Commit: 9c4ccae4f4c3c7f383cbb96955fa34faca92684b

**Comment by blvdmitry** [comment]
Tried running the tests locally and the test that errored out has passed. Is there a chance one of the e2e tests can be flaky? I've pressed "update branch" to see if the outcome in the CI would be different (haven't expected it will need another approval, but makes sense 😅). Otherwise, that might be my local env and I'll try to make it fail the same way it does in CI here

**Comment by samcx** [comment]
@blvdmitry It needs another approval!

Started discussing this internally. I'd love to ask some questions around this!

- What do you think about keeping the default fallbacks (one of the big benefits of using `next/font`) hashed still? I assume if the 3rd-parties are just targeting, e.g., Inter, keeping the default fallbacks hashed seems to make sense.
- Also curious what 3rd party libraries you are attempting to use with `next/font`!

**Comment by blvdmitry** [comment]
I'm scratching my own itch here trying to have it supported for reshaped.so, but before picking it up I saw that the same issue exists for others too. This stands for any component library that provides pre-built themes that include custom fonts, as well as other 3rd party tools like builder.io mentioned in the original discussion.

Let me think about the fallbacks hashing, I wrote this code 1.5 years ago so don't remember all the details right now. I think that was either about fallbacks being not useful with hashes, considering that the 3rd party css for that font is already generated or it was actually breaking something during rendering when using 3rd party libraries. 

Let me check it in the local environment again with and without fallback hashing

**Comment by blvdmitry** [comment]
@samcx Reading the code again, I'm thinking about the following: 

The use case I have myself and also mentioned by others is when you already have a CSS pre-compiled with font-family css rules or css variables containing the font-family value with the fallbacks. For example, in Reshaped, you're getting the following variables when building a theme output:

image

So I think current PR implementation might be a bit confusing because in the use cases - fonts are used differently. Currently, next/font generates `@font-face` rules as well as a className for those rules. With the `fixedFontFamily`, CSS for the className (or any other selector) is already provided by the 3rd party library, so we would care only about the part responsible for building the `@font-face` rules with good defaults and building/loading the assets themselves. We don't actually need to use the className returned by the `Inter` call. 

Because of that:

- If someone is using `fixedFontFamily`, I would expect both font itself and its fallback to not have the hash added. Hashed fallback still won't be usable but having it without the hash lets developers rely on the fallback name in 3rd party tooling if needed.
- I should probably drop all the logic related to the className hashing updated in this PR, it shouldn't really matter for this problem
- What we can pay attention to instead is making it possible to call the font function without expecting it to assign its return value to a variable. Right now the loader throws an error in this case.

wdyt about these items? If you think this is a good approach, I'll update the first 2 items tomorrow and can look further into the loader errors after that

**Comment by blvdmitry** [comment]
Just checking, does this approach sound reasonable for you, folks?

**Comment by samcx** [comment]
@blvdmitry Haven't been able to circle back with the rest of team on this yet! Will let you know when I have a confirmation.

**Comment by samcx** [comment]
> If someone is using fixedFontFamily, I would expect both font itself and its fallback to not have the hash added. Hashed fallback still won't be usable but having it without the hash lets developers rely on the fallback name in 3rd party tooling if needed.

> I should probably drop all the logic related to the className hashing updated in this PR, it shouldn't really matter for this problem

Sounds good for both!

> It allows developers to pick the fontFamily name that is going to be used in the CSS output instead of the default one and make it work with vendor CSS files.

@blvdmitry Can you clarify the reason for specifying a name? Not sure if this is necessary (would there be a case where the vendored CSS would have a custom name)?

**Comment by blvdmitry** [comment]
I couldn't think of a "valid" reason to call the fonts differently from their original name, however it's possible in the @font-face rule so it felt wrong ignoring that in the solution. So theoretically I can turn it into a boolean flag instead and just remove the hashing while acknowledging the trade-off of some 3rd party css files not being supported potentially. I don't have an example of such library at hand, maybe someone following the PR has? (or maybe not)

**Comment by blvdmitry** [comment]
@samcx I've updated the PR to not change the className hashing since it can be used fine even with the hashing. Tested in my local project too. I've also kept the option as a string value right now based on my comment above but I can drop it in case there are any concerns on your side.

The only thing I haven't addressed right now is the part when next-swc throws a build error whenever the function is not assigned to a variable right now (which we might no longer needed when used with `fixedFontFamily`). Since that part is written with Rust, it'll take me some time to figure out since I haven't used Rust before. Also if anyone from the team can help or anyone reading this would like to contribute – that would definitely let us get to the ready state faster. Meanwhile I'll keep reading the code 😅

**Comment by blvdmitry** [comment]
@samcx not sure what was happening in the failing pipelines. Couldn't reproduce it locally after merging in latest canary so pushed the updated branch again now

**Comment by samcx** [comment]
Sorry for the delayed responses!

> not sure what was happening in the failing pipelines. Couldn't reproduce it locally after merging in latest canary so pushed the updated branch again now

@blvdmitry Same thing as before, we have to either commit a commit ourself or manually approve your commit.

> "valid" reason to call the fonts differently from their original name, however it's possible in the @font-face rule so it felt wrong ignoring that in the solution

We weren't able to think of a reason to call, e.g., "Inter", something else than "Inter" so I think it's safe to just make this a `boolean` flag. Maybe have to rename this from `fixedFontFamily` if that's the case?

Let me know what you think!

**Comment by blvdmitry** [comment]
@samcx sure, no problem, I'm fine with either too. I will update the PR later this week

**Comment by samcx** [comment]
@blvdmitry Thanks for the update!

Sorry for this (I should have waited longer to comment on the PR so I don't possibly make you do extra work), but we've been talking more about this internally and we're not sure if we should even be hashing by default (e.g., would you need two different ways of doing Inter). Let me know what are your additional thoughts on that!

**Comment by blvdmitry** [comment]
Hm, no problem at all and as someone who has a use case of this – I would definitely be leaning towards saying yes, that can be the default 😅 

But I wonder if there is any history about the original decision to hash it. From outside that looks like a decision aligned with the general css hashing strategy where all generated parts are hashed (e.g. css modules, keyframe names, etc.). I can also think of a use case a next.js project installing a google font without hashing, while the user has a broken local version of the font installed which might be conflicting with the font provided by Next.js. 

Can't think of any other real use related to this but there might be something I'm missing when it comes to scoped theming in css

**Comment by samcx** [comment]
> looks like a decision aligned with the general css hashing strategy where all generated parts are hashed

I've discussed this with more folks within the team and the consensus of why we came to this decision is precisely this (unfortunately, the engineer who originally built most of next/font is no longer on our team).

We haven't been able to think of a strong reason to hash the `fontFamily` name/scope font usage, so we think we should be fine with this minor change to the behavior.

**Comment by blvdmitry** [comment]
Is there any process you have internally to validate these assumptions and avoid breaking changes? If not and it's usually assumption and past experience based – I'm happy to just remove the default hashing in the next commit

**Comment by samcx** [comment]
> Is there any process you have internally to validate these assumptions and avoid breaking changes?

The general process is starting a discussion with the anyone who was involved (e.g. in this case, Tim, Malte, getting the historical context) behind the design/decision-making, confirm with others with a lot of experience with the codebase, and ask others on the Next.js team.

**Comment by blvdmitry** [comment]
Updated the implementation to not use css hashing by default (classNames are still hashed).
Output example from my local project with Inter used together with Reshaped, works as expected

image

File: packages/next/src/build/webpack/loaders/next-font-loader/postcss-next-font.ts
Comment (blvdmitry) [reply packages/next/src/build/webpack/loaders/next-font-loader/postcss-next-font.ts:46]: It wraps the family with the quotes in case the font name itself is using any special characters that would break the css property assignment when used without quotes

**Comment by blvdmitry** [comment]
Hey @samcx, anything I can do/help with to get this PR moving futher?

**Comment by thomasbennedbaek** [comment]
@samcx Re: [your earlier comment](https://github.com/vercel/next.js/pull/53608#issuecomment-1984758725):

> We weren't able to think of a reason to call, e.g., "Inter", something else than "Inter" so I think it's safe to just make this a `boolean` flag. Maybe have to rename this from `fixedFontFamily` if that's the case?

What about e.g. "Open Sans", "Source Code Pro" and other fonts with spaces, hyphens, or other special characters in their name?

To my understanding, in the currently suggested implementation, the font-family name will be the name of the `const` you assigned the font loader to. E.g. `const OpenSans = Open_Sans(options)` is outputted as `font-family: 'OpenSans'` in the `@font-face` rule.

This becomes an issue when the 3rd party library (or, in our case, a decoupled UI) references the font by `Open Sans` (in two words) which would be an invalid `const` name.

In my opinion, that's a strong case for letting developers rename the font-family name in the CSS output, as @blvdmitry initially suggested 🙂

**Comment by dannyrb** [comment]
Hey! Reading up on the comments, it's not clear to me if anything is outstanding or if this is ready for review.

Happy to toss some dev time behind this if there are outstanding changes preventing review!

**Comment by blvdmitry** [comment]
@dannyrb Just need to check what's mentioned in https://github.com/vercel/next.js/pull/53608#issuecomment-2049779976 and get an approval

**Comment by timneutkens** [comment]
@blvdmitry I had a look at what Google fonts does by default, which is just `'Open Sans'`, as long as that matches I think it's fine? I.e. url: https://fonts.googleapis.com/css2?family=Open+Sans:ital,wght@0,300..800;1,300..800&display=swap

**Comment by blvdmitry** [comment]
@timneutkens Seems so. I would assume that it's just the generation script for all the exported fonts afterwards that replaces spaces with underscores?

**Comment by LouisCuvelierStadion** [comment]
Hey,
Thanks for this PR. We are currently stuck because of this missing feature. What's currently blocking ?

**Comment by jhopley-stadion** [comment]
I am facing an issue, where i need to explicitly  defined my font face name. Can this be reviewed and merged ?

**Comment by blvdmitry** [comment]
@samcx would you have some time next week to check if we need anything else before merging this? I think one of the pipelines was failing last time but I felt like I couldn't figure out the connection back then and with each rebase it has to be manually approved to run again, so will need your help resolving that last bit

**Comment by blvdmitry** [comment]
Thanks for updating the tests @ijjk, I was just looking through and updating the new once once that appeared since the last ci run and just noticed you've already pushed the changes 🙌

**Comment by asaroukou** [comment]
Would be great if shipped no the next release ❤️

**Review by ijjk — APPROVED** [review]

**Comment by ijjk** [comment]
Thanks for the PR!

**Pull request merged**
